### PR TITLE
Validate cipher updates with revision date

### LINKED ISSUE #780: [BUG] Change data in the web overwritten password saved from Chrome extension
### Subject of the issue
(Note: I'm not sure if this is a bug in the official Chrome extension or in the bitwraden_rs project. But I'm gonna report it here first.)

**This is a potential data loss bug that can be easily reproduced, so would appreciate if this can be prioritized.**

### Your environment
* Bitwarden_rs version: 1.13.0-4cec502f 
* Install method: Docker image
* Clients used: Chrome extension
* Reverse proxy and version: Pomerium (probably not related)
* Version of mysql/postgresql: 
* Other relevant information: 
 
### Steps to reproduce
1. Add a new entry in the vault. In my case, I created an entry with password `simple` with login URl `https://example.com`.
2. In the web, go to "Tools" then "Weak Passwords Report", you should find `example.com` in the list as `Very Weak`.
3. You click on that entry, you should be at the "EDIT ITEM" page. Keep it open.
4. Now click on the Chrome extension icon, find `example.com`, modify the password to `complex`, save. At this time, you should see "Password history: $NUM" in the Chrome extension.
5. To make sure the server received this change, in the extension, go to "Settings" -> "Sync" -> "Sync Vault Now".
6. Now go back to the web, the "EDIT ITEM" page, modify anything on the form (e.g. change username to `foo`) then save.
7. Go back to vault in the web, search example.com and check the password

### Expected behaviour
The username of the entry should be `foo` and the password should be `complex`, with at least one password history with value `simple`.

### Actual behaviour
The username is `foo` as expected, the password is `simple`, there's no password history at all. I.e. the password that I saved from the browser extension is gone.

FWIW: this is actually happening to me when I go through weak passwords report one by one and modify the password in a different tab, and accidentally saved the form with modified data in the web UI.

### PR DESCRIPTION
**Pull request by jjlin**
Prevent clients from updating a cipher if the local copy is stale.
Validation is only performed when the client provides its last known
revision date; this date isn't provided when using older clients,
or when the operation doesn't involve updating an existing cipher.

Fixes #780.

### CODE REVIEW

**Comment by jjlin** [comment]
This corresponds to changes in bitwarden/server#994. This feature hasn't been released upstream yet, but I've tested it with a build of their latest web vault code.

**Comment by BlackDex** [comment]
Looks good. I think this will resolve #780 right? At least when the clients are updated.

**Comment by jjlin** [comment]
More or less. There's still a 1 second window where the server would accept two conflicting writes, but that's not too likely. I think that window could be reduced to as small as 1 millisecond, but I'd have to see what time resolution the other clients provide.

File: src/api/core/ciphers.rs [hunk 1]
```diff
@@ -238,6 +247,20 @@ pub fn update_cipher_from_data(
     nt: &Notify,
     ut: UpdateType,
 ) -> EmptyResult {
+    // Check that the client isn't updating an existing cipher with stale data.
+    if data.LastKnownRevisionDate.is_some() {
+        let dt = data.LastKnownRevisionDate.unwrap();
+        let res = NaiveDateTime::parse_from_str(&dt, "%+"); // ISO 8601 format
+        if res.is_ok() {
+            let dt = res.unwrap();
+            if cipher.updated_at.signed_duration_since(dt).num_seconds() > 1 {
```
Comment (jjlin) [inline src/api/core/ciphers.rs:256]: BTW, the web vault seems to provide a datetime with millisecond resolution, whereas bitwarden_rs seems to maintain nanosecond resolution, so in principle this slop factor could be reduced to as little as 1 ms, assuming the other clients also provide millisecond resolution.

File: src/api/core/ciphers.rs [hunk 2]
```diff
@@ -238,6 +247,20 @@ pub fn update_cipher_from_data(
...
+                err!("The client copy of this cipher is out of date. Resync the client and try again.")
+            }
+        } else {
+            warn!("error parsing LastKnownRevisionDate '{}': {}", dt, res.unwrap_err());
+        }
+    }
```
Comment (cynecx) [inline src/api/core/ciphers.rs:262]: This is more idiomatic rust code:

```suggestion
    if let Some(dt) = &data.LastKnownRevisionDate {
        match NaiveDateTime::parse_from_str(dt, "%+") {
            Ok(dt) if cipher.updated_at.signed_duration_since(dt).num_seconds() > 1 => {
                err!("The client copy of this cipher is out of date. Resync the client and try again.");
            }
            Ok(_) => (),
            Err(err) => warn!("error parsing LastKnownRevisionDate '{}': {}", dt, err),
        }
    }
```

File: src/api/core/ciphers.rs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 use std::collections::{HashMap, HashSet};
 use std::path::Path;
 
+use chrono::{NaiveDateTime, Utc};
```
Comment (cynecx) [inline src/api/core/ciphers.rs:4]: `Utc` is unused.

Comment (BlackDex) [reply src/api/core/ciphers.rs:4]: @cynecx, i think you have to check again ;). See the `if soft_delete` part at the bottom of the page.

Comment (cynecx) [reply src/api/core/ciphers.rs:4]: Sorry, my bad. I've only applied the changes above so rustc put out a diagnostic.

File: src/api/core/ciphers.rs (hunk 2 above)
Comment (jjlin) [reply src/api/core/ciphers.rs:262]: Thanks, I wasn't aware of conditionals on match arms. I amended to use a variant of your suggestion.

**Comment by jjlin** [comment]
Actually, it turns out client support is already included in v2.17.1 of the web vault, but hasn't made it into a mobile client release yet. So this PR would be functional in conjunction with dani-garcia/bw_web_builds#21.

File: src/api/core/ciphers.rs (hunk 1 above)
Comment (dani-garcia) [reply src/api/core/ciphers.rs]: Yeah I think the chances of a modification in a 1 second timespan are slim enough that 1 second will do fine.

**Comment by dani-garcia** [comment]
The web vault with the changes is building now, once it's done I'll change the dockerfiles to point to it, but in the meantime this can be merged already.

**Pull request merged**